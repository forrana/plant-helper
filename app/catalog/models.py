from django.db import models
from django.contrib.postgres.fields import ArrayField
import datetime

class PlantEntry(models.Model):
    nick_name = models.CharField(max_length=200)
    scientific_name = models.CharField(max_length=200)
    nick_names = ArrayField(models.CharField(max_length=200, blank=True), null=True)
    days_between_watering_growing = models.DurationField(default=datetime.timedelta(days=5))
    days_between_watering_dormant = models.DurationField(default=datetime.timedelta(days=10))


    def __str__(self):
        return f"{self.scientific_name}.{self.nick_name}"

    def save(self, *args, **kwargs):
        is_new = True if not self.pk else False
        super(PlantEntry, self).save(*args, **kwargs)
        if is_new:
            for nick_name in self.nick_names:
                print(nick_name)
                nick_name_entry = NickName(plant_entry=self, name=nick_name)
                nick_name_entry.save()


    @property
    def days_between_watering_growing_int(self):
        return self.days_between_watering_growing.days

    @property
    def days_between_watering_dormant_int(self):
        return self.days_between_watering_dormant.days

    class Meta:
        indexes = [models.Index(fields=['nick_name']), models.Index(fields=['scientific_name'])]

class NickName(models.Model):
    plant_entry = models.ForeignKey(PlantEntry, on_delete=models.CASCADE)
    name = models.CharField(max_length=256)

    def __str__(self):
        return f"{self.name}"


    class Meta:
        indexes = [models.Index(fields=['name'])]
