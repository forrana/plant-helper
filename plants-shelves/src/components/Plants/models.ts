export interface PlantType {
  id: number;
  name: string;
  scientificName: string;
  daysUntilNextWatering: number;
  daysBetweenWatering: number;
  daysBetweenWateringGrowing: number;
  daysBetweenWateringDormant: number;
  symbol: SymbolType;
  daysPostpone: number;
}

export interface SymbolType {
  userWideId: number
}

export interface PlantsData {
  plants: PlantType[];
}

export interface Subscription {
  endpoint: string,
  p256dh: string,
  auth: string,
  permission_given: boolean
}

export interface SubscriptionData {
  subscription: boolean
}

export interface PlantData {
  plant: PlantType;
}

export interface PlantSuggestion {
  nickName: string,
  nickNames: Array<string>,
  scientificName: string,
  daysBetweenWateringGrowingInt: number,
  daysBetweenWateringDormantInt: number
}

export interface PlantNickName {
  name: string,
  plantEntry: PlantSuggestion
}

export interface GlobalState extends PlantsData { }

export interface loadAction { type: 'load', plants: PlantType[] }
export interface addAction { type: 'add', plant: PlantType }
export interface updateAction { type: 'update', plant: PlantType, index: number }
export interface deleteAction { type: 'delete', index: number }

export type GlobalReducerAction =
  | addAction
  | loadAction
  | updateAction
  | deleteAction;
