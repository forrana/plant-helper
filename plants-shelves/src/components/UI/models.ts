export interface MessageType { description: string, color: "danger" | "info" | "warning" | "success" }

export interface AddMessage  { type: 'addMessage', message: MessageType }
export interface RemoveMessage  { type: 'removeMessage' }

export interface AlertState { messages: MessageType[] }

export type AlertReducerAction =
  | AddMessage
  | RemoveMessage