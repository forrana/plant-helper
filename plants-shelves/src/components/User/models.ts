export interface LoginError {
  message: string;
  code: string;
}
