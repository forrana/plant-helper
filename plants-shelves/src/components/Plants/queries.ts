import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
    symbol {
      userWideId
    }
    room {
      id,
      roomName,
      colorBackground
    }
  }
}
`;

const GET_ALL_ROOMS = gql`
  query {
    rooms {
      id, roomName, colorBackground
    }
  }
`;
const ROOMS_BY_NAME_FRAGMENT = gql`
  query RoomsByNameFragment($nameFragment: String!) {
    roomsByNameFragment(nameFragment: $nameFragment) {
      id, roomName, colorBackground
    }
  }
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $groupName: String, $colorBackground: String) {
      createPlant(plantName: $plantName, scientificName: $scientificName, daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
        daysBetweenWateringDormant: $daysBetweenWateringDormant, groupName: $groupName, colorBackground: $colorBackground) {
        plant {
          id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
          symbol {
            userWideId
          }
          room {
            id,
            roomName,
            colorBackground
          }
        }
      }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const POSTPONE_WATERING = gql`
  mutation PostponeWatering($plantId: ID!, $postponeDays: Int!) {
    postponeWatering(plantId: $plantId, days: $postponeDays) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_PLANT = gql`
  mutation UpdatePlant($plantId: ID!, $plantName: String!,
    $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $postponeDays: Int!, $groupName: String, $colorBackground: String) {
    updatePlant(plantId: $plantId, plantName: $plantName, scientificName: $scientificName,
      daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
      daysBetweenWateringDormant: $daysBetweenWateringDormant,
      postponeDays: $postponeDays,
      groupName: $groupName,
      colorBackground: $colorBackground) {
      plant {
        id, name, scientificName, daysUntilNextWatering,
        daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant,
        daysPostpone,
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_ROOM = gql`
  mutation UpdateRoom($roomId: ID!, $roomName: String!) {
    updateRoom(roomId: $roomId, roomName: $roomName) {
      ok
      room {
        id,
        roomName,
        colorBackground
      }
    }
  }
`

const CREATE_ROOM = gql`
  mutation CreateRoom($plant1Id: ID!, $plant2Id: ID!) {
    createRoom(plant1Id: $plant1Id, plant2Id: $plant2Id) {
      ok
    }
  }
`;

const DELETE_PLANT = gql`
  mutation DeletePlant($plantId: ID!) {
    deletePlant(plantId: $plantId) {
      ok
    }
  }
`;

const GET_SUBSCRIPTION = gql`
  query {
    subscription
  }
`

const CREATE_SUBSCRIPTION = gql`
  mutation CreateSubscription($endpoint: String!, $p256dh: String!, $auth: String!, $permissionGiven: Boolean!) {
    createSubscription(endpoint: $endpoint, p256dh: $p256dh, auth: $auth, permissionGiven: $permissionGiven) {
      ok
    }
  }
`

export {
  GET_ALL_PLANTS,
  GET_ALL_ROOMS,
  ADD_PLANT,
  WATER_PLANT,
  UPDATE_PLANT,
  DELETE_PLANT,
  CREATE_SUBSCRIPTION,
  GET_SUBSCRIPTION,
  POSTPONE_WATERING,
  CREATE_ROOM,
  ROOMS_BY_NAME_FRAGMENT,
  UPDATE_ROOM
}
