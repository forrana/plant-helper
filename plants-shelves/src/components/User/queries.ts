import { gql } from '@apollo/client';

const CREATE_USER = gql`
  mutation Register($email: String!, $username: String!, $password1: String!, $password2: String!) {
    register(email: $email, username: $username, password1: $password1, password2: $password2) {
      success,
      errors
    }
  }
`;

const LOG_IN = gql`
  mutation TokenAuth($username: String!, $password: String!) {
    tokenAuth(username: $username, password: $password) {
      success,
      errors,
      unarchiving,
      token,
      refreshToken,
      unarchiving,
      user {
        id,
        username,
      }
    }
  }
`;


const REFRESH_TOKEN = gql`
  mutation RefreshToken($refreshToken: String!) {
    refreshToken(
      refreshToken: $refreshToken
    ) {
      success,
      errors,
      payload,
      token,
      refreshToken,
    }
  }
`

const UPSERT_USER_SETTINGS = gql`
  mutation UpsertUserSettings($startTime: String!, $endTime: String!, $timezone: String) {
    upsertUserSettings(startTime: $startTime, endTime: $endTime, timezone: $timezone) {
      ok
    }
  }
`

const GET_USER_SETTINGS = gql`
  query {
    userSettings {
      notificationsStartTime, notificationsEndTime, timezone
    }
  }
`;

export { CREATE_USER, LOG_IN, REFRESH_TOKEN, UPSERT_USER_SETTINGS, GET_USER_SETTINGS }
