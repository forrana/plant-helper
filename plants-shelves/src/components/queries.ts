import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
  }
}
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!) {
    createPlant(plantName: $plantName, scientificName: $scientificName) {
      plant {
        name,
        scientificName
      }
    }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
      }
    }
  }
`;


export { GET_ALL_PLANTS, ADD_PLANT, WATER_PLANT }
