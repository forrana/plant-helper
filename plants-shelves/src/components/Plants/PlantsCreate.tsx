import React, { useState, useContext } from 'react';
import { Redirect } from "react-router-dom";
import { useMutation } from '@apollo/client';
import { Button, Form, FormGroup, Label, Input, Spinner } from 'reactstrap';

import { ADD_PLANT } from './queries'
import PlantsDispatch from './PlantsDispatch';
import { PlantData } from './models'
import AutoCompleteInput from '../UI/AutoCompleteInput';

interface PlantsCreateProps { action?: () => any }

function PlantsCreate({ action }: PlantsCreateProps) {
  const dispatch = useContext(PlantsDispatch);
  const [submitted, setSubmitted] = useState(false);
  const [plantName, setPlantName] = useState("");
  const [scientificName, setScientificName] = useState("");

  const [addPlant, { loading, error }] = useMutation(ADD_PLANT, {
    onCompleted: (data: { createPlant: PlantData }) => {
      dispatch && dispatch({ type: 'add', plant: data.createPlant.plant })
      setPlantName("");
      setScientificName("");
      setSubmitted(true);
      action && action();
    }
  });

  const handlePlantNameInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setPlantName(event.target.value);
  };

  const handleScientificNameInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setScientificName(event.target.value);
  };

  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if(plantName && scientificName) {
      addPlant({ variables: { plantName, scientificName } })
    }
  }

  if (submitted) return <Redirect push to="/"/>

  if (loading) return  <Spinner color="primary" />

  if (error)  return  <p>Error :( {error.message}</p>;

  return (
    <Form
      onSubmit={handleFormSubmit}
      autocomplete="off"
    >
      <FormGroup>
        <Label for="name">Name:</Label>
        <Input type="text" name="name" id="name" placeholder="Plant name"
          value={plantName}
          onChange={handlePlantNameInputChange}
        />
      </FormGroup>
      <FormGroup>
        <Label for="scientificName">Scientific name:</Label>
        <AutoCompleteInput
          type="text" name="scientificName" id="scientificName" placeholder="Scientific name"
          value={scientificName}
          onChange={handleScientificNameInputChange}
        />
      </FormGroup>
      <Button type="submit">Add plant</Button>
    </Form>
  )
}

export { PlantsCreate }
