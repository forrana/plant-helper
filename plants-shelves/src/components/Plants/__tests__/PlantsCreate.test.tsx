import React from 'react';
import { act, fireEvent, render, screen, Screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import { GlobalState, PlantType } from '../models';
import { ADD_PLANT } from '../queries';
import { PlantsCreate } from '../PlantsCreate';
import { MemoryRouter } from 'react-router-dom';

const newPlant: PlantType = {
  id: 1,
  name: "Aloe 1",
  scientificName: "Aloe Vera 1" ,
  daysUntilNextWatering: 0,
  daysBetweenWatering: 7
 };

const mocksWithPlant: any = [
  {
    request: {
      query: ADD_PLANT,
      variables: { plantName: newPlant.name, scientificName: newPlant.scientificName}
    },
    result: {
      data:  newPlant,
    },
  },
  {
    request: {
      query: ADD_PLANT,
      variables: { plantName: "error", scientificName: newPlant.scientificName}
    },
    error:
      { message: "Something went wrong" },
  },
];

const mockCallback = jest.fn(() => null)

const createPlant = async (screen: Screen) => {
  await act(async () => {
    fireEvent.change(
      screen.getByTestId("name-input"),
      { target: { value: newPlant.name } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByTestId("sc-name-input"),
      { target: { value: newPlant.scientificName } }
    )
  });

  await act(async () => {
    fireEvent.click(
      screen.getByText(/Add plant/i),
      { target: { value: newPlant.scientificName } }
    )
    await new Promise(resolve => setTimeout(resolve, 0));
  });
}

test('Match snapshot', async () => {
  const {container} = render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithPlant} addTypename={false}>
        <PlantsCreate action={mockCallback}/>
      </MockedProvider>
    </MemoryRouter>,
  );

  expect(container).toMatchSnapshot()
});

test('Allow to create plant', async () => {
 render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithPlant} addTypename={false}>
        <PlantsCreate action={mockCallback}/>
      </MockedProvider>
    </MemoryRouter>,
  );

  await act(async () => {
    fireEvent.change(
      screen.getByTestId("name-input"),
      { target: { value: newPlant.name } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByTestId("sc-name-input"),
      { target: { value: newPlant.scientificName } }
    )
  });

  await act(async () => {
    fireEvent.click(
      screen.getByText(/Add plant/i),
      { target: { value: newPlant.scientificName } }
    )
    await new Promise(resolve => setTimeout(resolve, 0));
  });

  expect(mockCallback).toBeCalledTimes(1)
});

test('Don\'t create plant with empty fields', async () => {
  render(
     <MemoryRouter>
       <MockedProvider mocks={mocksWithPlant} addTypename={false}>
         <PlantsCreate action={mockCallback}/>
       </MockedProvider>
     </MemoryRouter>,
   );

   await act(async () => {
     fireEvent.click(
       screen.getByText(/Add plant/i),
       { target: { value: newPlant.scientificName } }
     )
     await new Promise(resolve => setTimeout(resolve, 0));
   });

   expect(mockCallback).not.toBeCalled()
 });

 test('Don\'t create plant with error response', async () => {
  render(
     <MemoryRouter>
       <MockedProvider mocks={mocksWithPlant} addTypename={false}>
         <PlantsCreate action={mockCallback}/>
       </MockedProvider>
     </MemoryRouter>,
   );

   await act(async () => {
     fireEvent.click(
       screen.getByText(/Add plant/i),
       { target: { value: newPlant.scientificName } }
     )
     await new Promise(resolve => setTimeout(resolve, 0));
   });

   await act(async () => {
    fireEvent.change(
      screen.getByTestId("name-input"),
      { target: { value: "error" } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByTestId("sc-name-input"),
      { target: { value: newPlant.scientificName } }
    )
  });

  await act(async () => {
    fireEvent.click(
      screen.getByText(/Add plant/i),
      { target: { value: newPlant.scientificName } }
    )
    await new Promise(resolve => setTimeout(resolve, 0));
  });

   expect(mockCallback).not.toBeCalled()
 });