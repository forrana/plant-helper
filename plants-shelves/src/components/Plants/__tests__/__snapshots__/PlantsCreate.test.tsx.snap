// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
<div>
  <form
    autocomplete="off"
    class=""
    data-testid="plant-create-form"
  >
    <div
      class="form-group"
    >
      <label
        class=""
        for="name"
      >
        Name:
      </label>
      <input
        class="form-control"
        data-testid="name-input"
        id="name"
        name="name"
        placeholder="Plant name"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class=""
        for="scientificName"
      >
        Scientific name:
      </label>
      <input
        class="form-control"
        data-testid="sc-name-input"
        id="scientificName"
        name="scientificName"
        placeholder="Scientific name"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <section
        class="label"
      >
        <span
          class="icon-droplet"
        />
        <h4
          class="header"
        >
          <span
            class="icon-droplet"
          />
        </h4>
      </section>
      <input
        class="form-control-range"
        data-testid="days-between-watering-input"
        id="daysBetweenWatering"
        max="30"
        min="1"
        name="daysBetweenWatering"
        placeholder="Days between watering"
        required=""
        type="range"
        value="7"
      />
      <div
        class="text-center"
      >
        <small>
          7
           day(s) between waterings
        </small>
      </div>
    </div>
    <button
      class="button btn btn-secondary"
      type="submit"
    >
      Add plant
    </button>
  </form>
</div>
`;
